src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -29,9 +31,34 @@ internal sealed partial class SerialStream : Stream
         private int _writeTimeout = 0;
         private byte[] _tempBuf = new byte[1];
         private Task _ioLoop;
+        private object _ioLoopLock = new object();
         private ConcurrentQueue<SerialStreamIORequest> _readQueue = new ConcurrentQueue<SerialStreamIORequest>();
         private ConcurrentQueue<SerialStreamIORequest> _writeQueue = new ConcurrentQueue<SerialStreamIORequest>();
 
+        private long _totalBytesRead = 0;
+        private long TotalBytesAvailable => _totalBytesRead + BytesToRead;
+        private long _lastTotalBytesAvailable;
+
+        // called when one character is received.
+        private event SerialDataReceivedEventHandler _dataReceived;
+        internal event SerialDataReceivedEventHandler DataReceived
+        {
+            add
+            {
+                bool wasNull = _dataReceived == null;
+                _dataReceived += value;
+
+                if (wasNull)
+                {
+                    EnsureIOLoopRunning();
+                }
+            }
+            remove
+            {
+                _dataReceived -= value;
```

Comment:
Can you stop the loop if _dataReceived becomes null?

Reply:
it does stop after 2s if there are no new read/write requests